### ISSUE TITLE
Gvisor fails to run after b80cba61: platform/ptrace: make some operations arch specific #1248

### ISSUE THREAD

**Issue by majek** [google/gvisor#1408]
Pulling out discussion under https://github.com/google/gvisor/pull/1248 merge: https://github.com/google/gvisor/pull/1248#issuecomment-566218835

After that merge I can't run gvisor. I see:
```
$ sudo runsc --debug --debug-log=/tmp/xyz/ --net-raw create hello
creating container: waiting for sandbox to start: EOF
```
and see cash in dmesg:
```
[293673.381030] exe[14539]: segfault at 0 ip 000000000042f3e9 sp 00007fff6736da70 error 6 in runsc[400000+9ef000]
[293673.381056] Code: 08 48 89 44 24 10 48 8b 44 24 30 48 89 44 24 18 48 8d 44 24 38 48 89 44 24 20 48 8d 44 24 08 48 89 04 24 e8 c9 e8 02 00 31 c0 <48> c7 00 00 00 00 00 48 8b 6c 24 28 48 83 c4 30 c3 cc cc cc cc cc
```
(before that merge, I all works fine)

The debug log:
```
$ tail /tmp/runsc/runsc.log.20191217-084129.170041.boot
...
D1217 07:41:29.201828   14521 x:0] syscall filter getcpu: [] => 0x7472617020283029
W1217 07:41:29.338894   14521 x:0] unable to get registers: no such process
panic: ptrace status unexpected: got 139, wanted stopped
stubStart        = 00007fffffff0000


goroutine 35 [running]:
gvisor.dev/gvisor/pkg/sentry/platform/ptrace.(*thread).dumpAndPanic(0xc000232000, 0xc00010a3c0, 0x50)
        pkg/sentry/platform/ptrace/subprocess.go:327 +0x3cc
gvisor.dev/gvisor/pkg/sentry/platform/ptrace.(*thread).wait(0xc000232000, 0x0, 0x0)
        pkg/sentry/platform/ptrace/subprocess.go:364 +0x484
gvisor.dev/gvisor/pkg/sentry/platform/ptrace.attachedThread(0x0, 0x50000, 0x0, 0x0, 0x0)
        pkg/sentry/platform/ptrace/subprocess_linux.go:216 +0xd45
gvisor.dev/gvisor/pkg/sentry/platform/ptrace.probeSeccomp(0x0)
        pkg/sentry/platform/ptrace/subprocess_linux.go:42 +0x66
gvisor.dev/gvisor/pkg/sentry/platform/ptrace.createStub(0x0, 0x0, 0x0)
        pkg/sentry/platform/ptrace/subprocess_linux.go:96 +0x3f
gvisor.dev/gvisor/pkg/sentry/platform/ptrace.newSubprocess.func1(0x103f1c8, 0xc0000960c0, 0xc000096120)
        pkg/sentry/platform/ptrace/subprocess.go:163 +0x87
created by gvisor.dev/gvisor/pkg/sentry/platform/ptrace.newSubprocess
        pkg/sentry/platform/ptrace/subprocess.go:158 +0x196

goroutine 1 [chan receive]:
gvisor.dev/gvisor/pkg/sentry/platform/ptrace.newSubprocess(0x103f1c8, 0x0, 0x0, 0x0)
        pkg/sentry/platform/ptrace/subprocess.go:204 +0x1bc
gvisor.dev/gvisor/pkg/sentry/platform/ptrace.New.func1()
        pkg/sentry/platform/ptrace/ptrace.go:194 +0x58
sync.(*Once).doSlow(0x1783ff8, 0x103f1c0)
        GOROOT/src/sync/once.go:66 +0xe3
sync.(*Once).Do(0x1783ff8, 0x103f1c0)
        GOROOT/src/sync/once.go:57 +0x45
gvisor.dev/gvisor/pkg/sentry/platform/ptrace.New(0x0, 0x0, 0x0)
        pkg/sentry/platform/ptrace/ptrace.go:188 +0x4a
gvisor.dev/gvisor/pkg/sentry/platform/ptrace.(*constructor).New(0x1783f80, 0x0, 0x0, 0x0, 0x0, 0x0)
        pkg/sentry/platform/ptrace/ptrace.go:244 +0x47
gvisor.dev/gvisor/runsc/boot.createPlatform(0xc0001b3440, 0x0, 0x0, 0x0, 0x0, 0x0)
        runsc/boot/loader.go:402 +0x305
gvisor.dev/gvisor/runsc/boot.New(0x7fff6736efe3, 0x5, 0xc0001a70a0, 0xc0001b3440, 0x4, 0x0, 0xc0001ee060, 0x1, 0x1, 0xc00009fa60, ...)
        runsc/boot/loader.go:188 +0x325
gvisor.dev/gvisor/runsc/cmd.(*Boot).Execute(0xc0001061b0, 0x11604a0, 0xc0000a8000, 0xc0001da5a0, 0xc000186d60, 0x2, 0x2, 0x0)
        runsc/cmd/boot.go:226 +0x1225
github.com/google/subcommands.(*Commander).Execute(0xc0000ca000, 0x11604a0, 0xc0000a8000, 0xc000186d60, 0x2, 0x2, 0x0)
        external/com_github_google_subcommands/subcommands.go:200 +0x515
github.com/google/subcommands.Execute(0x11604a0, 0xc0000a8000, 0xc000186d60, 0x2, 0x2, 0x0)
        external/com_github_google_subcommands/subcommands.go:481 +0x68
main.main()
        runsc/main.go:316 +0x293b

goroutine 34 [syscall]:
os/signal.signal_recv(0x0)
        GOROOT/src/runtime/sigqueue.go:147 +0x9c
os/signal.loop()
        GOROOT/src/os/signal/signal_unix.go:23 +0x22
created by os/signal.init.0
        GOROOT/src/os/signal/signal_unix.go:29 +0x41
(END)
```

For the record, I'm running rather fresh Linux kernel `Linux version 5.4.0-rc5+` on AMD64.

Full debug logs:

[runsc.log.20191217-084129.158932.create.txt](https://github.com/google/gvisor/files/3972006/runsc.log.20191217-084129.158932.create.txt)
[runsc.log.20191217-084129.160287.gofer.txt](https://github.com/google/gvisor/files/3972007/runsc.log.20191217-084129.160287.gofer.txt)
[runsc.log.20191217-084129.170041.boot.txt](https://github.com/google/gvisor/files/3972008/runsc.log.20191217-084129.170041.boot.txt)

Calling in @xiaobo55x and @fvoznika

**Comment by prattmic** [comment]
Could you disassemble the runsc binary and see which function the crash IP (0x42f3e9) is in?

```
$ objdump -d path/to/runsc | less
```

Search for `42f3e9:`.

I suspect the problem is that enableCpuidFault is not marked nosplit, in which case I'd expect the crash to be either there in in a Go runtime function.

**Comment by prattmic** [comment]
Please try https://github.com/google/gvisor/commit/78d8bb98288b539b853e46130205208d9176b0bd and let me know if that fixes the problem for you.

FWIW, my Go compiler is inlining enableCpuidFault, so while the nosplit directive needs to be there for correctness, it isn't actually making a difference in my build. If it is the same for you then this won't fix the problem.

**Comment by xiaobo55x** [comment]
Hi @prattmic, @majek, @fvoznika,

I have verified the #78d8bb9 commit, and it can fix this issue.

During preparing the #1248 I only verified the fast-build mode(default) which enable almost all the compiler optimizations(including inline).  When it come to dbg mode, the inline was disabled, and the
enableCpuidFault() would be compiled to a function with a preamble to manage a goroutine’s stack, which would cause this issue. 

Sorry for the inconvenience！

Regards,
Haibo
